src/estimagic/optimization/subsolvers/_trsbox_quadratic.py

```diff
@@ -0,0 +1,624 @@
+
+        beta = gradient_sumsq / gradient_sumsq_old
+        continue
+
+    if need_alt_trust_step:
+        curve_min = 0
+        x_candidate = _perform_alternative_trustregion_step(
+            x_candidate=x_candidate,
+            x_bounded=x_bounded,
+            gradient_candidate=gradient_candidate,
+            model_hessian=model_hessian,
+            lower_bounds=lower_bounds,
+            upper_bounds=upper_bounds,
+            n_fixed_variables=n_fixed_variables,
+            total_reduction=total_reduction,
+        )
+    else:
+        x_candidate = _apply_bounds_to_candidate_vector(
+            x_candidate, x_bounded, lower_bounds, upper_bounds
+        )
+
+    return x_candidate
+
+
+def _perform_alternative_trustregion_step(
+    x_candidate,
+    x_bounded,
+    gradient_candidate,
+    model_hessian,
+    lower_bounds,
+    upper_bounds,
+    n_fixed_variables,
+    total_reduction,
+):
+    """Perform the alternative trust-region step."""
+    n = len(x_candidate)
+    max_iter = 100 * n**2
+
+    for _ in range(max_iter):
+        if n_fixed_variables >= n - 1:
+            x_candidate = _apply_bounds_to_candidate_vector(
+                x_candidate, x_bounded, lower_bounds, upper_bounds
+            )
+            break
+
+        search_direction = np.zeros(n)
+        search_direction[x_bounded == 0] = x_candidate[x_bounded == 0]
+
+        x_reduced = _calc_sum_sq(x_candidate[x_bounded == 0])
+        x_grad = x_candidate[x_bounded == 0].T @ gradient_candidate[x_bounded == 0]
+        gradient_reduced = _calc_sum_sq(gradient_candidate[x_bounded == 0])
+        hess_s = model_hessian @ search_direction
+        hessian_reduced = hess_s
+
+        restart_alt_loop = False
+
+        for _ in range(max_iter):
+            raw_reduction = gradient_reduced * x_reduced - x_grad**2
+            if raw_reduction <= 1.0e-4 * total_reduction**2:
+                restart_alt_loop = False
+                break
+
+            search_direction, s_norm = _compute_new_search_direction_and_norm(
+                x_candidate,
+                x_bounded,
+                x_reduced,
+                gradient_candidate,
+                x_grad,
+                raw_reduction,
+            )
+
+            (
+                x_bounded,
+                index_active_bound,
+                n_fixed_variables,
+                active_bound,
+                bound_on_tangent,
+                free_variable_reached_bound,
+            ) = _calc_upper_bound_on_tangent(
+                x_candidate,
+                search_direction,
+                x_bounded,
+                lower_bounds,
+                upper_bounds,
+                n_fixed_variables,
+            )
+
+            if free_variable_reached_bound:
+                restart_alt_loop = True
+                break
+
+            hess_s = model_hessian @ search_direction
+
+            s_hess_s = np.sum(search_direction[x_bounded == 0] * hess_s[x_bounded == 0])
```

Comment:
Lines ``261``- ``265`` shoould be mathematically equivalent to

```python
s_hess_s = search_direction[x_bounded == 0].T @ hess_s[x_bounded == 
x_hess_s = x_candidate[x_bounded == 0].T @ hess_s[x_bounded == 0]
x_hess_x = x_candidate[x_bounded == 0].T @ hessian_reduced[x_bounded == 0]
```
Numerically, however, there are slight differences in the ``more_wild`` benchmarks, where the ``@`` version performs slightly worse.